File: src/main/java/be/breina/openrgb/model/MatrixMap.java
package be.breina.openrgb.model;

import static lombok.AccessLevel.PRIVATE;

import java.io.IOException;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;
import be.breina.openrgb.util.LittleEndianInputStream;

@ToString
@NoArgsConstructor(access = PRIVATE)
@Getter
public class MatrixMap {

    private int height;
    private int width;
    private int[][] matrix;

    MatrixMap(LittleEndianInputStream input) throws IOException {
        height = input.readInt();
        width = input.readInt();

        matrix = new int[height][width];

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                matrix[i][j] = input.readInt();
            }
        }
    }
}

File: src/main/kotlin/be/breina/openrgb/model/MatrixMap.kt
package be.breina.openrgb.model

import be.breina.openrgb.util.LittleEndianInputStream

class MatrixMap private constructor(val height: Int, val width: Int, val matrix: Array<IntArray>) {

    internal companion object {
        fun from(input: LittleEndianInputStream): MatrixMap? {
            if (input.readUnsignedShort() <= 0) {
                return null;
            }

            val height: Int = input.readInt()
            val width: Int = input.readInt()
            val matrix: Array<IntArray> = Array(height) { IntArray(width) { input.readInt() } }

            return MatrixMap(height, width, matrix)
        }
    }
}